### PR TITLE
feat(manager): Adds Unity3D

### PR DESCRIPTION
**Pull request by ViMaSter**
## Changes

Adds a manager for updating the runtime of a project using the Unity game engine.

## Context

See #27219, closes #27295.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
  - Ran tests locally and via PR on my fork: https://github.com/ViMaSter/renovate/pull/7

### CODE REVIEW

File: lib/modules/manager/unity3d/index.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+export { extractPackageFile } from './extract';
+
+export const defaultConfig = {
+  fileMatch: fileMatchRegex,
+  packageRules: [
+    {
+      groupName: 'Unity Editor',
+      matchPackageNames: ['m_EditorVersion', 'm_EditorVersionWithRevision'],
+    },
+  ],
```
Comment (secustor) [inline lib/modules/manager/unity3d/index.ts:13]: This should be rather part of a preset, as it is hard to find out what is happening if the config is set here. 

If we are talking about the same version in different places you can set `depName` to the same value. This will by default result too in a single PR

File: lib/modules/manager/unity3d/extract.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  if (!matches) {
+    return null;
+  }
+  const key = matches.groups?.depName;
+  const version = matches.groups?.currentValue;
+  if (!key || !version || !supportedExtensionsKey.includes(key)) {
+    return null;
+  }
+
+  if (version !== version.trim()) {
```
Comment (secustor) [inline lib/modules/manager/unity3d/extract.ts:23]: Why is this necessary? 🤔

File: lib/modules/manager/unity3d/extract.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+export const fileMatchRegex: string[] = [
+  '(^|/)ProjectSettings/ProjectVersion.txt',
+];
+
+const supportedExtensionsKey = [
+  'm_EditorVersionWithRevision',
+  'm_EditorVersion',
+];
+const parseLine = (line: string): PackageDependency | null => {
+  const matches = regEx(/^(?<depName>.+): (?<currentValue>.+)/g).exec(line);
```
Comment (secustor) [inline lib/modules/manager/unity3d/extract.ts:13]: Move the regex definition to the module scope.

File: lib/modules/manager/unity3d/extract.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+export const fileMatchRegex: string[] = [
+  '(^|/)ProjectSettings/ProjectVersion.txt',
+];
+
+const supportedExtensionsKey = [
+  'm_EditorVersionWithRevision',
+  'm_EditorVersion',
+];
+const parseLine = (line: string): PackageDependency | null => {
```
Comment (secustor) [inline lib/modules/manager/unity3d/extract.ts:12]:
```suggestion
function parseLine(line: string): PackageDependency | null {
```
No unnecessary anonymous functions.

File: lib/modules/manager/unity3d/extract.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  };
+};
+
+export function extractPackageFile(
+  content: string,
+  fileName: string,
+): PackageFileContent {
+  if (!fileMatchRegex.every((regex: string) => regEx(regex).test(fileName))) {
+    return { deps: [] };
+  }
```
Comment (secustor) [inline lib/modules/manager/unity3d/extract.ts:45]: This isn't necessary, `extractPackageFile` will only called on fitting file names

File: lib/modules/manager/unity3d/extract.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  return {
+    autoReplaceStringTemplate:
+      '{{depName}}{{#if newValue}}:{{newValue}}{{/if}}',
+    currentDigest: undefined,
+    currentValue: version,
+    datasource: 'unity3d',
+    depName: key,
+    depType: 'final',
+    replaceString: matches[0],
```
Comment (secustor) [inline lib/modules/manager/unity3d/extract.ts:35]:
```suggestion
    currentValue: version,
    datasource: 'unity3d',
    depName: key,
    depType: 'final',
```
These vales are set by default. 

Also why `final` as `depType`, if there is only one?

File: lib/modules/manager/unity3d/extract.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  return {
+    deps: content
+      .split('\n')
+      .map((line) => line.trim())
+      .filter((line) => line.length > 0)
+      .map(parseLine)
+      .filter((dep) => dep !== null) as PackageDependency[],
+  };
```
Comment (secustor) [inline lib/modules/manager/unity3d/extract.ts:54]: Chaining these makes it hard to debug it please use a for each loop. 
https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#iterating-objects--containers

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/unity3d/extract.ts (hunk 7 above)
Comment (viceice) [reply lib/modules/manager/unity3d/extract.ts:54]: and return null for empty deps

**Review by viceice — COMMENTED** [review]
☝️

File: lib/constants/category.ts [hunk 8]
```diff
@@ -24,6 +24,7 @@ const Categories = [
   'rust',
   'swift',
   'terraform',
+  'unity3d',
```
[the remark this answers is not in the document]
Comment (ViMaSter) [reply lib/constants/category.ts:27]: Unity has [a different package manager (Unity Package Manager)](https://docs.unity3d.com/Manual/upm-concepts.html) for its sub-dependencies and while I've read #19300, I haven't got it working using the proposed NPM workaround: There's a lock file different from normal NPM lock files that would have to be generated and currently isn't. This might be something worth adding under the `unity3d` umbrella.

Let me know if you find any existing ones more appropriate.

[the remark this answers is not in the document]
Comment (secustor) [reply lib/constants/category.ts:27]: Makes sense to add the category then, if support for this manager is added too. 

Please remove it for now.

File: lib/modules/manager/unity3d/extract.spec.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+          datasource: 'unity3d',
+          depName: 'm_EditorVersionWithRevision',
+          depType: 'final',
+          replaceString:
+            'm_EditorVersionWithRevision: 2021.3.35f1 (157b46ce122a)',
+        },
+      ]);
+    });
+
+    it('skips if casing is incorrect', () => {
```
Comment (HonkingGoose) [inline lib/modules/manager/unity3d/extract.spec.ts:143]: Should Renovate raise a warning about the malformed casing? Or should Unity warn about malformed files?

File: lib/modules/manager/unity3d/extract.spec.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+      const res = extractPackageFile(
+        'm_EditorVersionWITHRevision: 2022.3.19f1 (244b723c30a6)\n',
+        'ProjectSettings/ProjectVersion.txt',
+      )?.deps;
+      expect(res).toBeEmpty();
+    });
+
+    it('skips abnormal spacing', () => {
```
Comment (HonkingGoose) [inline lib/modules/manager/unity3d/extract.spec.ts:151]: Same question about Renovate warning or Unity warning here.

File: lib/modules/manager/unity3d/readme.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+Handles version numbers inside `ProjectSettings/ProjectVersion.txt` files used by the Unity game engine.
```
Comment (HonkingGoose) [inline lib/modules/manager/unity3d/readme.md:1]: I see two ways to read this:

- Renovate can read `ProjectSettings.txt` and `ProjectVersion.txt` files
- Renovate can read a `ProjectVersion.txt` file _if_ its in the `ProjectSettings` directory

I don't know which interpretation is correct, as I don't use Unity.

Please update the text so it explains precisely what Renovate can update. 🙂

File: lib/modules/manager/unity3d/readme.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+`ProjectSettings/ProjectVersion.txt` always contains two version references in a yml-like syntax; one with and one without a hash, both of which are updated:
```
Comment (HonkingGoose) [inline lib/modules/manager/unity3d/readme.md:3]:
```suggestion
`ProjectSettings/ProjectVersion.txt` always have two version references in a yml-like syntax; one with and one without a hash, both are updated:
```

Also, is it YML-like syntax or YAML-syntax? I lean towards YAML-syntax, but not sure. 🙂

File: lib/modules/manager/unity3d/readme.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+
+```yml
+m_EditorVersion: 2020.3.15f2
+m_EditorVersionWithRevision: 2020.3.15f2 (6cf78cb77498)
+```
+
+Relies on the corresponding `unity3d` datasource and versioning.
```
Comment (HonkingGoose) [inline lib/modules/manager/unity3d/readme.md:10]: I don't see any changes to a datasource directory in this PR? Maybe that was added in a different PR? Or I'm misunderstanding something here. 🙃

In any case, please check if this sentence is correct.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@ViMaSter do you have time to address the PR review comments?

**Comment by bdovaz** [comment]
Any way to help speed up this PR?

The next step after this PR is to support Unity package updates with another manager.

**Comment by secustor** [comment]
The requested changes have to implemented before we can consider merging it.    
As `ViMaSter` seems to be occupied, you can take over the implementation.

**Comment by rarkins** [comment]
Closing due to inactivity. Reopening or replacing this PR in future would be welcomed
